### ISSUE TITLE
How to install local NPM module and it's dependencies to project?

### ISSUE THREAD

**Issue by edouard-lopez** [npm/npm#18228]
#### I'm opening this issue because:

  - [ ] npm is crashing.
  - [ ] npm is producing an incorrect install.
  - [x] npm is doing something I don't understand.
  - [ ] Other (_see below for feature requests_):

#### What's going wrong?

I'm working on shareable `eslint` configuration to spread automation and internal best-practices. So I'm adding some eslint plugins and tools as dependencies.

### Env

```
$ lsb_release -a
No LSB modules are available.
Distributor ID: LinuxMint
Description:    Linux Mint 18.2 Sonya
Release:        18.2
Codename:       sonya
```

    $ node --version; 
    v8.2.1

    $ npm --version
    5.3.0

### Local package

I created a local `npm` project, says `eslint-config-company`, with the following `package.json`:

    {
      "name": "eslint-config-company",
      "version": "0.0.1",
      "main": "index.js",
      "dependencies": {
        "eslint": "4.x.x",
        "eslint-plugin-ember-suave": "1.x.x",
        "eslint-plugin-prettier": "2.x.x",
        "prettier": "1.x.x"
      },
      "files": [
        "index.js"
      ],
      "keywords": [
        "eslint",
        "eslintconfig"
      ],
      "repository": "company/eslint-config-company",
      "private": true
    }

and files:

    index.js  package.json  README.md  yarn.lock

### Installing

Then, I install my `eslint-config-company` package into a another project to test it:

    npm install --save-dev --verbose ../eslint-config-company

In `node_modules/` directory I got the `eslint-config-company/` as a symlink:

    $ ls node_modules/eslint-config-company -lah
    lrwxrwxrwx 1 me me 29 Aug 17 22:02 node_modules/eslint-config-company -> ../../eslint-config-company/

but no trace of either `prettier` nor `eslint-plugin-prettier`

    $ ls node_modules/{prettier,eslint-plugin-prettier}
    ls: cannot access 'node_modules/prettier': No such file or directory
    ls: cannot access 'node_modules/eslint-plugin-prettier': No such file or directory

N.B.: `eslint` and `eslint-plugin-ember-suave` are already dependencies of that project thus exists.

## Question

* Am I wrong assuming that my package's `dependencies` will install on `npm install` ?
* How do I fix my package in order to install them?

**related:** [How to install local NPM module and it's dependencies to project?](https://stackoverflow.com/q/45744183/802365)

**Comment by emememw** [comment]
Can confirm. This definitely worked before v5. What **does** work is publishing the module (`eslint-config-company`) and installing it directly from the registry. Seems like this problem only exists when installing modules from local directories.

**Comment by benschenker** [comment]
Do you have a `package-lock.json` in your `eslint-config-company` project?  My understanding is that if it does, npm will install the dependencies scoped just to `eslint-config-company`.  You can confirm this is the case if those modules are available in `node_modules/eslint-config-company/node_modules`.

**Comment by blackconcerto** [comment]
I got the same problem,too. Before v5, I can build my javascript lib with package.json, and install it in the other project. However,  it can't work with v5, the project with the package.json config "lib:file:../lib" can't not include the lib.

**Comment by Butterfly** [comment]
Hi, @edouard-lopez I've not had a chance to do something like what you're trying, but I was looking at your file paths there. With this command, you appear to be telling it to install up one directory above:

`npm install --save-dev --verbose ../eslint-config-company`

When you run the `ls` (list directory contents) command, it then shows your target as being *two* directories above:

> `ls node_modules/eslint-config-company -lah`
lrwxrwxrwx 1 me me 29 Aug 17 22:02 node_modules/eslint-config-company -> ../../eslint-config-company/

* Is that where it's supposed to be, i.e. two directories above the symlink? That would put it exiting upward two layers and then pointing back in from whatever is the parent directory above node_modules.

* If you only move up one directory, is that where you find the missing `prettier` and `eslint-plugin-prettier` that you know exist? Or even if you don't find them there, do you find them where they should be if that symlink was only pointing up one directory instead of two?

To save time since I'm not always at the keyboard, if that target is only supposed to be one directory above where `ls` says it's sitting, maybe it's about [basic/default configurations](https://docs.npmjs.com/cli/config) for installation paths? If that's not it, I'm in over my head after that. :smile:

**Comment by Butterfly** [comment]
I had thought of something else, but like I said, I'm over my head a little here. Very coincidentally, @KenanY [left a recent comment](https://github.com/npm/npm/issues/18391#issuecomment-327019180) on #18391 where they said exactly what I was thinking here.

If directory levels (hierarchy) was one of the problems and that gets fixed, I was also wondering about how it works with the same name being used for different things. @KenanY 's suggestion is to make some small change that distinguishes the various symlinks, directories, etc, from each other's names. I'm seeing that suggestion come up within a few issues and also about a few places where it might matter, e.g. the data we [enter in `package.json`](https://github.com/npm/npm/issues/18310#issuecomment-326766511). Best wishes :smile:

**Comment by oieduardorabelo** [comment]
I think this is also related to: https://github.com/npm/npm/issues/18418

**Comment by mitsos1os** [comment]
Yes this is the same problem... this happened in node 8... up to node 7.10.1 it works correctly

**Comment by fbrbovic** [comment]
Any update on this, I have the same problem
